### ISSUE TITLE
UseStaticGraphEvaluation results in item not being expanded inside property

### ISSUE THREAD

**Issue by ViktorHofer** [NuGet/Home#9398]
In dotnet/runtime we use the [property item conversion feature](https://docs.microsoft.com/en-us/visualstudio/msbuild/comparing-properties-and-items?view=vs-2019#conversions-between-strings-and-item-lists): https://github.com/dotnet/runtime/blob/master/src/installer/pkg/projects/Directory.Build.props#L166. When using the new NuGet static graph evaluation mode, that property doesn't get expanded. It works fine when setting `/p:RestoreUseStaticGraphEvaluation=false`.

See the generated projects.assets.json below:

```
{
  "version": 3,
  "targets": {
    ".NETCoreApp,Version=v5.0": {
     ...
    },
    ".NETCoreApp,Version=v5.0/@(RestoreBuildRID)": {
      ...
    },
```

@jeffkl

**Comment by jeffkl** [comment]
This is by design at the moment, the static graph restore is not executing targets but the normal restore is.  Specifically, this `RuntimeIdentifiers` is actually being evaluated [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L706).

**Comment by ViktorHofer** [comment]
But that property is set during evaluation outside of a target. Are you saying that the property item conversion is running as part of a target?

**Comment by jeffkl** [comment]
Yeah during top-level evaluation, the property is not fully evaluated:
![image](https://user-images.githubusercontent.com/17556515/78583261-74463700-77eb-11ea-928c-d94212472e13.png)
![image](https://user-images.githubusercontent.com/17556515/78583300-81fbbc80-77eb-11ea-8a16-37f760fdc7b3.png)

But if you run a target, it gets fully evaluated:
![image](https://user-images.githubusercontent.com/17556515/78583347-95a72300-77eb-11ea-973c-dcc11b2d6233.png)

Example project:
```xml
<Project>
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <RestoreBuildRID Include="win-x64;linux-x64" />
  </ItemGroup>

  <PropertyGroup>
    <RuntimeIdentifiers>@(RestoreBuildRID)</RuntimeIdentifiers>
  </PropertyGroup>

  <Target Name="Foo">
    <ItemGroup>
      <Foo Include="$([System.Guid]::NewGuid())"
           RuntimeIdentifiers="$(RuntimeIdentifiers);$(RuntimeIdentifier)" />
    </ItemGroup>
  </Target>
</Project>
```

**Comment by ViktorHofer** [comment]
Hmm that's unfortunate as it means property item conversion isn't supported at all in a restore context. Any idea how to support this scenario Jeff? I suppose we aren't the only ones that use this feature.

**Comment by jeffkl** [comment]
Technically some targets are run before the property is read, so you can leverage that:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <RestoreBuildRID Include="win-x64;linux-x64" />
  </ItemGroup>

  <Target Name="SetRuntimeIdentifiersForRestore"
          BeforeTargets="CollectPackageReferences">
    <PropertyGroup>
      <RuntimeIdentifiers>@(RestoreBuildRID)</RuntimeIdentifiers>
    </PropertyGroup>
  </Target>
</Project>
```

```json
"runtimes": {
  "linux-x64": {
    "#import": []
  },
  "win-x64": {
    "#import": []
  }
}
    
```

**Comment by ViktorHofer** [comment]
Thanks Jeff for the workaround which I just validated and it works. Is there a long term solution planned? I expect this will impact more people going as they try out that feature.

**Comment by jeffkl** [comment]
The overwhelming majority of users have very simple logic to specify their runtimes, frameworks, and packages. The logic that was broken here is very complex and I doubt very many users will have the issue. Unfortunately restore was running targets before as a matter of implementation and the side effect is that people have taken a dependency on it. But the only way to speed up restore is to do less. This is why the feature is opt-in as some users will have to change their logic to adopt it. I’m just hoping the improvement is so dramatic that users will be willing to do the work.

**Comment by ViktorHofer** [comment]
Makes sense. Thanks for sharing your perspective.
